### DISCUSSION TITLE
aaccff24-99ff-4c96-b7e4-30f852d4c16d

### DISCUSSION THREAD

**Discussion by giscus** [varofla/blog_giscus#17] [General]
# aaccff24-99ff-4c96-b7e4-30f852d4c16d

MCU에 전원이 전원이 공급된 후 main 함수는 어떻게 실행될까요? MCU의 깊고 깊은 어딘가를 파보았습니다.

https://www.varofla.com/aaccff24-99ff-4c96-b7e4-30f852d4c16d

**Comment by mong87suk** [comment]
boot가 main flsh mode 되어 있을 때 STM32 마이크로컨트롤러에서 0x8000000에서 부터 정의된 영역만 사용하만 될 것 같은데
왜 0x8000000 부터 정의된 코드를 0x0000000 에 복사하는 이유가 뭔가요?

**Reply by varofla** [reply]
Memory Mapping은 무언가를 저장하기 위한것이 아니기 때문에 복사(copy) 보다는 연결(aliased)이 더 알맞은 표현인것 같습니다.
Flash의 주소는 0x0800 0000 인 반면, Cortex-M3 프로세서는 0x0 주소부터 실행되어, Flash의 벡터 테이블을 프로세서에 넣어주기 위해서는 이걸 0x0 위치에 연결해야 하죠.
데이터시트에 언급되었듯이 boot 모드에 따라 유저 어플리케이션이 실행되는 main flash 모드에서는 flash의 주소인 0x0800 0000에 연결하거나, 내장 부트로더의 경우 0x1FFF F000에 연결됩니다.

**Reply by mong87suk** [reply]
네 말씀하신 내용이 맞는 것 같습니다.
아래 빨간줄로 체크해 주신 내용들 다시 확인해보니 Flash memory에 있는 내용을 0x0 주소로 aliased 해주네요.
다시 확인 해주셔서 감사합니다.
Boot from main Flash memory: the main Flash memory is aliased in the boot memory
space (0x0000 0000), but still accessible from its original memory space (0x800 0000). 

저는 pc의 address를 0x04로 설정 시에도 정상 동작하는지 아래와 같이 확인해 보았습니다.
내용 확인 부탁 드립니다.

STM32F103 시리즈 보드로 아래와 같이 테스트를 진행해 보았습니다.

.ld 파일 링크 스크립터 파일에 FLASH 주소를 아래와 같이 설정 한 후 빌드하여 .bin 파일을 반든 후
FLASH    (rx)    : ORIGIN = 0x0000000,   LENGTH = 64K

STM32CubeProgrammer를 이용하여 Download를 진행해 보았습니다.

.bin 파일 확인 시
0x0000000: 20005000 
0x0000004: 00000399

STM32CubeProgrammer를 이용하여 Download를 진행 후 Disconnect 후 
STM32CubeProgrammer로 다시 Connect 후 CPU 항목에서 PC 레지스터 값을 확인 결과
PC: 0x0000398

위와 같이 확인 결과 정상적으로 실행이 되었습니다.

하지만 STM32CubeIDE로는 디버깅이 되지 않는 현상을 발견 했습니다.

FLASH    (rx)    : ORIGIN = 0x0000000 로 설정 시에도 정상 동작은 하지만
FLASH    (rx)    : ORIGIN = 0x8000000으로 하는 이유는 디버깅을 하기 위해서 설정한 것이 아닌가 생각이 됩니다.
제가 생각하는게 맞는지 확인 부탁 드립니다.

또 하나 궁금증이 있습니다.

memory aliased 방식은 st의 boot 메카니즘으로 flash memory 값을 0x0으로 mapping 한 후
sp loading 한 뒤에 pc 값을 증가시켜 pipeline(fetch, decode, execute ..)이 동작 되도록 하는 설계 구조 인가요?

**Reply by varofla** [reply]
재미나고 심오한 질문 남겨주셔서 감사합니다 :laughing: 

정황상으로는 CubeProgrammer에 연결되기 이전에 main flash로 부팅되어서 0x0이 0x0800 0000에 aliased된 상태이기 때문에 펌웨어가 정상적으로 올라간 반면, 디버깅 모드에서는 aliased되는 위치가 다르던가 한 것이 아닌가 생각합니다.
또는 디버깅 모드에서는 CPU 레지스터를 직접 주무를 수 있으니 실제 Flash의 데이터를 Fetch하는 대신 디버거를 통해 직접 명령을 넣어주는게 아닌가? 하는 생각도 들고요.

aliased 타이밍과 관련된 내용은 데이터시트를 통해 확인할 수 있습니다.
[RM0008 레퍼런스 메뉴얼](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwiJx5Cxu9iEAxWIplYBHdBmAqAQFnoECAYQAQ&url=https%3A%2F%2Fwww.st.com%2Fresource%2Fen%2Freference_manual%2Frm0008-stm32f101xx-stm32f102xx-stm32f103xx-stm32f105xx-and-stm32f107xx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf&usg=AOvVaw0gKEwqSurS9WnLC3KmOqn8&opi=89978449)의 61페이지 두 번째 문단 내용을 확인해보면 이런 내용이 있습니다.
> The BOOT pins are also re-sampled when exiting from Standby mode. Consequently they must be kept in the required Boot mode configuration in Standby mode.<ins>After this startup delay has elapsed, the CPU fetches the top-of-stack value from address 0x0000 0000, then starts code execution from the boot memory starting from 0x0000 0004.</ins>

즉, flash 등이 aliased된 후 프로세서의 파이프라인이 돌아갑니다.

아쉽게도 첫 질문에 대해서는 지금 제 밑천으로는 삽질 없이 정답을 얻어낼 수 없을것 같아, 조금의 삽질을 해보고 해답이 보이면 답글 드리도록 하겠습니다 :+1:

**Comment by chochmun** [comment]
좋은글 감사합니다.
